sv2/handlers-sv2/src/template_distribution.rs

```diff
@@ -169,46 +271,92 @@ pub trait HandleTemplateDistributionMessagesFromServerAsync {
 /// Whether this is relevant or not depends on which object is implementing the trait, and whether
 /// this contextual information is readily available or not. In cases where `client_id` is either
 /// irrelevant or can be inferred without the context, this should always be `None`.
+///
+/// ## TLV Extension Support
+///
+/// The `tlv_data` parameter in message handlers contains validated TLV fields if the message has
+/// extension data appended. TLV fields are only passed if they match negotiated extensions
+/// returned by `get_negotiated_extensions_with_client()`.
 pub trait HandleTemplateDistributionMessagesFromClientSync {
     type Error: HandlerErrorType;
 
+    /// Returns the list of negotiated extension_types with a client.
+    ///
+    /// Used to validate TLV fields appended to messages. Return an empty Vec if no
+    /// extensions have been negotiated.
+    fn get_negotiated_extensions_with_client(&self, client_id: Option<usize>) -> Vec<u16>;
+
+    /// Handles a raw Template Distribution protocol message frame from a client.
+    ///
+    /// This method parses the raw frame, extracts any TLV extension data, and delegates
+    /// to `handle_template_distribution_message_from_client` with the parsed message and TLV fields.
     fn handle_template_distribution_message_frame_from_client(
         &mut self,
         client_id: Option<usize>,
-        message_type: u8,
+        header: Header,
         payload: &mut [u8],
     ) -> Result<(), Self::Error> {
-        let parsed: TemplateDistribution<'_> = (message_type, payload)
-            .try_into()
-            .map_err(Self::Error::parse_error)?;
-        self.handle_template_distribution_message_from_client(client_id, parsed)
+        let negotiated_extensions = self.get_negotiated_extensions_with_client(client_id);
+        if negotiated_extensions.is_empty() {
+            let parsed: TemplateDistribution<'_> = (header.msg_type(), payload)
+                .try_into()
+                .map_err(Self::Error::parse_error)?;
+            self.handle_template_distribution_message_from_client(client_id, parsed, None)
+        } else {
```

Comment:
Lets return early, and remove this else.
```suggestion
           return self.handle_template_distribution_message_from_client(client_id, parsed, None);
        } 
```